kernel/src/log_segment.rs

```diff
@@ -112,13 +126,245 @@ impl LogSegment {
+        self
+    }
+    /// Build the [`LogSegment`]
+    ///
+    /// This fetches checkpoint and commit files using the `fs_client`.
+    pub(crate) fn build(self) -> DeltaResult<LogSegment> {
+        if self.start_version.is_some() && self.start_checkpoint.is_some() {
+            return Err(Error::generic("Failed to build LogSegment: Cannot specify both start_version and start_checkpoint"));
+        }
+        let Self {
+            fs_client,
+            table_root,
+            start_checkpoint,
+            start_version,
+            end_version,
+            commit_files_sorted_ascending,
+            omit_checkpoint_parts,
+        } = self;
+        let log_root = table_root.join("_delta_log/").unwrap();
+        let (mut sorted_commit_files, mut checkpoint_parts) = match (start_checkpoint, end_version)
+        {
+            (Some(cp), None) => Self::list_log_files_with_checkpoint(&cp, fs_client, &log_root)?,
+            (Some(cp), Some(version)) if cp.version <= version => {
+                Self::list_log_files_with_checkpoint(&cp, fs_client, &log_root)?
+            }
+            _ => Self::list_log_files_from_version(fs_client, &log_root, None)?,
+        };
+
+        if omit_checkpoint_parts {
+            checkpoint_parts.clear();
+        }
+
+        // Commit file versions must satisfy the following:
+        // - Be greater than or equal to the start version
+        // - Be greater than the most recent checkpoint version if it exists
+        // - Be less than or equal to the end version.
+        if let Some(start_version) = start_version {
+            sorted_commit_files.retain(|log_path| log_path.version >= start_version);
+        }
+        if let Some(checkpoint_file) = checkpoint_parts.first() {
+            sorted_commit_files.retain(|log_path| checkpoint_file.version < log_path.version);
+        }
+        if let Some(end_version) = end_version {
+            sorted_commit_files.retain(|log_path| log_path.version <= end_version);
+        }
+
+        // get the effective version from chosen files
+        let version_eff = sorted_commit_files
+            .last()
+            .or(checkpoint_parts.first())
+            .ok_or(Error::MissingVersion)? // TODO: A more descriptive error
+            .version;
+        if let Some(end_version) = end_version {
+            require!(
+                version_eff == end_version,
+                Error::MissingVersion // TODO more descriptive error
+            );
+        }
+
+        // We assume commit files are sorted in ascending order. If `commit_files_sorted_ascending`
+        // is false, reverse to make it descending.
+        if !commit_files_sorted_ascending {
+            sorted_commit_files.reverse();
+        }
+
+        Ok(LogSegment {
+            end_version: version_eff,
+            log_root,
+            commit_files: sorted_commit_files,
+            checkpoint_parts,
+        })
+    }
+
+    pub(crate) fn list_log_files_from_version(
```

Comment:
I'm unclear why this method and the one below are in the `impl` for `LogSegmentBuilder`? I guess you want to call them as `Self::xxx` above, but I don't think that gives any benefit, and I'd just move them out to stand-alone methods